design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,597 @@
+
+2. A user invokes an operation like `s3::GetObject`. A [params object is created](#creating-params). In the body
+   of `make_operation()`, this is passed to `config.endpoint_resolver` to load a Smithy-native endpoint. The Smithy
+   native endpoint is written into the property bag.
+3. The AWS endpoint middleware (non-generic) reads the endpoint out of the property bag and uses it to create an AWS
+   endpoint. See [converting to AWS Endpoint](#converting-a-smithy-endpoint-to-an-aws-endpoint).
+4. The request is dispatched!
+
+The other major piece of implementation required is actually implementing the rules engine. To learn more about
+rules-engine internals, skip to [implementing the rules engine](#implementing-the-rules-engine).
+
+### Code generating client context params
+
+When a smithy model uses the `@clientContextParams` trait, we need to generate client params onto the Rust SDK. This is
+a **Smithy-native** feature. This should be implemented as a "standard" config decorator that reads traits from the
+current
+model.
+
+<details>
+<summary>Kotlin Snippet for Client context params</summary>
+
+```kotlin
+class ClientContextDecorator(ctx: ClientCodegenContext) : NamedSectionGenerator<ServiceConfig>() {
+    private val contextParams = ctx.serviceShape.getTrait<ClientContextParamsTrait>()?.parameters.orEmpty().toList()
+        .map { (key, value) -> ContextParam.fromClientParam(key, value, ctx.symbolProvider) }
+
+    data class ContextParam(val name: String, val type: Symbol, val docs: String?) {
+        companion object {
+            private fun toSymbol(shapeType: ShapeType, symbolProvider: RustSymbolProvider): Symbol =
+                symbolProvider.toSymbol(
+                    when (shapeType) {
+                        ShapeType.STRING -> StringShape.builder().id("smithy.api#String").build()
+                        ShapeType.BOOLEAN -> BooleanShape.builder().id("smithy.api#Boolean").build()
+                        else -> TODO("unsupported type")
+                    }
+                )
+
+            fun fromClientParam(
+                name: String,
+                definition: ClientContextParamDefinition,
+                symbolProvider: RustSymbolProvider
+            ): ContextParam {
+                return ContextParam(
+                    RustReservedWords.escapeIfNeeded(name.toSnakeCase()),
+                    toSymbol(definition.type, symbolProvider),
+                    definition.documentation.orNull()
+                )
+            }
+        }
+    }
+
+    override fun section(section: ServiceConfig): Writable {
+        return when (section) {
+            is ServiceConfig.ConfigStruct -> writable {
+                contextParams.forEach { param ->
+                    rust("pub (crate) ${param.name}: #T,", param.type.makeOptional())
+                }
+            }
+            ServiceConfig.ConfigImpl -> emptySection
```

Comment:
If this is the smithy native generator, I think we should make the config field private and add a accessor `fn` to the `ConfigImpl` section